### PR TITLE
[Inliner] Add argument/function attribute propagation before inlining.

### PR DESCRIPTION
**Pull request by goldsteinn**
- [Inliner] Propagate callee function memory access attributes before inlining
- [Inliner] Propagate callee argument memory access attributes before inlining

### CODE REVIEW

**Comment by goldsteinn** [comment]
Note on compile time affect: https://llvm-compile-time-tracker.com/compare.php?from=2da4960f20f7e5d88a68ce25636a895284dc66d8&to=f8c9f5bce65756598da22e8aec5d91fb66b16d5c&stat=instructions%3Au

Minimal for normal O3, a bit more significant for LTO.

**Comment by goldsteinn** [comment]
ping.

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 1]
```diff
@@ -1374,6 +1376,129 @@ static AttrBuilder IdentifyValidPoisonGeneratingAttributes(CallBase &CB) {
...
+  }
+
+  // Won't be able to propagate anything.
+  if (!HasAttrToPropagate)
+    return;
+
+  for (BasicBlock &BB : *CalledFunction) {
+    for (Instruction &Ins : BB) {
+      CallBase *InnerCB = dyn_cast<CallBase>(&Ins);
+      if (InnerCB != nullptr) {
```
Comment (RKSimon) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:1410]: `if (auto *InnerCB = dyn_cast<CallBase>(&Ins))`

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 2]
```diff
@@ -1374,6 +1376,129 @@ static AttrBuilder IdentifyValidPoisonGeneratingAttributes(CallBase &CB) {
...
+        if (auto *NewInnerCB =
+                dyn_cast_or_null<CallBase>(VMap.lookup(InnerCB))) {
+          AttributeList AL = NewInnerCB->getAttributes();
+          for (unsigned I = 0, E = InnerCB->arg_size(); I < E; ++I) {
+            // Check if the underlying value for the parameter is an argument.
+            const Value *UnderlyingV =
+                getUnderlyingObject(InnerCB->getArgOperand(I));
+            if (const Argument *Arg = dyn_cast<Argument>(UnderlyingV)) {
```
Comment (RKSimon) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:1418]: `if (auto *Arg = dyn_cast<Argument>(UnderlyingV)) {`

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 3]
```diff
@@ -1374,6 +1376,129 @@ static AttrBuilder IdentifyValidPoisonGeneratingAttributes(CallBase &CB) {
...
+      assert(BBInfo != FirstAllocaAndCBs->end());
+      BBInfo->second.first = true;
+      // We have an alloca in a preceding BB, we can't propagate any memory
+      // effects.
+      return BBInfo->second;
+    }
+  }
+
+  auto BBInfo = FirstAllocaAndCBs->find(BB);
+  assert(BBInfo != FirstAllocaAndCBs->end());
```
Comment (RKSimon) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:1457]: (style) assert message

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 4]
```diff
@@ -1374,6 +1376,129 @@ static AttrBuilder IdentifyValidPoisonGeneratingAttributes(CallBase &CB) {
...
+    MemoryEffects ME) {
+  auto InsertRes = FirstAllocaAndCBs->insert({BB, {false, {}}});
+  if (!InsertRes.second)
+    return InsertRes.first->second;
+
+  for (BasicBlock *PBB : predecessors(BB)) {
+    auto PBBInfo = GetBBAllocaAndCallsiteInfo(PBB, FirstAllocaAndCBs, ME);
+    if (PBBInfo.first) {
+      auto BBInfo = FirstAllocaAndCBs->find(BB);
+      assert(BBInfo != FirstAllocaAndCBs->end());
```
Comment (RKSimon) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:1448]: (style) assert message

File: llvm/lib/Transforms/Utils/InlineFunction.cpp [hunk 5]
```diff
@@ -1374,6 +1376,129 @@ static AttrBuilder IdentifyValidPoisonGeneratingAttributes(CallBase &CB) {
...
+  DenseMap<BasicBlock *, std::pair<bool, SmallVector<CallBase *, 4>>>
+      FirstAllocaAndCBs;
+
+  for (BasicBlock &BB : *CalledFunction) {
+    auto BBInfo = GetBBAllocaAndCallsiteInfo(&BB, &FirstAllocaAndCBs, ME);
+    // We found no callsites that we can propagate memory effects to.
+    if (BBInfo.second.empty())
+      continue;
+    for (CallBase *OtherCB : BBInfo.second) {
+      assert(OtherCB->getParent() == &BB);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Utils/InlineFunction.cpp:1488]: (style) assert message

**Comment by goldsteinn** [comment]
Dropping this.

**Pull request closed without merging**